lib/Dialect/TritonGPU/Transforms/AMDReorderInstructions.cpp

```diff
@@ -0,0 +1,397 @@
+
+#define GEN_PASS_CLASSES
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
+
+using namespace mlir;
+namespace tt = triton;
+namespace ttg = triton::gpu;
+
+class TritonAMDGPUReorderInstructionsPass
+    : public TritonAMDGPUReorderInstructionsBase<
+          TritonAMDGPUReorderInstructionsPass> {
+public:
+  TritonAMDGPUReorderInstructionsPass() = default;
+
+  void sortOperandsByDominance(OperandRange operands,
+                               SmallVector<Value> &operandsSorted) {
+    ModuleOp m = getOperation();
+    mlir::DominanceInfo dom(m);
+
+    for (auto operand : operands) {
+      operandsSorted.push_back(operand);
+    }
+
+    if (operandsSorted.size() == 1) {
+      return;
+    }
+
+    std::sort(operandsSorted.begin(), operandsSorted.end(),
+              [&](const Value &a, const Value &b) {
+                Operation *operandA = a.getDefiningOp();
+                Operation *operandB = b.getDefiningOp();
+                if (operandA && operandB) {
+                  return dom.dominates(operandA, operandB);
+                }
+                return false;
+              });
+  }
+
+  void moveAfter(Operation *lhs, Operation *rhs) {
+    auto lhsId = getWSRoleId(lhs);
+    auto rhsId = getWSRoleId(rhs);
+    if (lhsId == rhsId)
+      lhs->moveAfter(rhs);
+  }
+
+  void moveBefore(Operation *lhs, Operation *rhs) {
+    auto lhsId = getWSRoleId(lhs);
+    auto rhsId = getWSRoleId(rhs);
+    if (lhsId == rhsId)
+      lhs->moveBefore(rhs);
+  }
+
+  bool isFAChainDot(tt::DotOp &dotOp) const {
+    SetVector<Operation *> slices;
+    getForwardSlice((Operation *)dotOp, &slices);
+
+    for (Operation *op : slices) {
+      if (isa<tt::DotOp>(op) && (op != dotOp)) {
+        auto operandA = op->getOperand(0).getDefiningOp();
+        auto containsOperandA =
+            std::find(slices.begin(), slices.end(), operandA) != slices.end();
+        if (containsOperandA) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  void moveImmediatelyAfterOperands(Operation *op,
+                                    SmallVector<Operation *> &movedOperations) {
+
+    if (std::find(movedOperations.begin(), movedOperations.end(), op) !=
+        movedOperations.end()) {
+      return;
+    }
+    auto operands = op->getOperands();
+    if (operands.empty()) {
+      return;
+    }
+    ModuleOp m = getOperation();
+    mlir::DominanceInfo dom(m);
+
+    for (auto operandVal : operands) {
+      Operation *argOp = operandVal.getDefiningOp();
+      if (!argOp) {
+        continue;
+      }
+      moveImmediatelyAfterOperands(argOp, movedOperations);
+    }
+
+    SmallVector<Value> operandsSorted;
+    sortOperandsByDominance(operands, operandsSorted);
+
+    if (!operandsSorted.empty() &&
+        operandsSorted[operandsSorted.size() - 1].getDefiningOp()) {
+
+      moveAfter(op, operandsSorted[operandsSorted.size() - 1].getDefiningOp());
+      if (failed(mlir::verify(m))) {
+        assert(false);
```

Comment:
Maybe simple assert?
```suggestion
      assert(succeeded(mlir::verify(m)));
```